### ISSUE TITLE
'Chronological' output within a section

### ISSUE THREAD

**Issue by fractalien** [spencermountain/wtf_wikipedia#88]
The parse method nicely yields arrays of sentences and links, but I see no way of knowing which lists were at which position. Is this loss of information deliberate, am I missing something, or was this just never an issue for anybody? A simple solution would seem to allow disabling the parsing of lists to somehow include them in the sentences.

**Comment by spencermountain** [comment]
Good point, maybe these should all be objects in a chronological array.
On vacation, will think about this when im back

**Comment by niebert** [comment]
I assume it is more tree with children in a chronlogical order
wiki article is the root node in an Abstract Syntax Tree AST
The wiki article is split up in top level sections ==My Section Title ==
* Before first section there could be a paragraph 
* All sections have a title wrapped with == and children, which are page elements.
* The sections are split into subsection and so forth that are children of the section node
* All subsections have a title wrapped with === and children, which are page elements.
* page elements are sentences, images links
* enumerations have arrays of items and the item are arrays of page elements.
arrays of page elements preserve the chronological output and nested enumerations can be handled properly for code generation.

**Comment by niebert** [comment]
Started to propose a solution for the order problem of content element see [GitHub Wiki](https://github.com/spencermountain/wtf_wikipedia/wiki/Paragraph-Recognition)

**Comment by spencermountain** [comment]
hey, this ordering-within-sections situation has been improved now with paragraph support.
Unfortunately, (and to after considerable work) I learned how all templates, references, and tables can have an infinite number of newlines within them. This means that they will remain on `Section` for the medium-term.
Images, lists, and sentences are now stored on Paragraph, so will now support a closer-to-truth ordering.

**Comment by spencermountain** [comment]
all the output methods will work normally, and should provide some better ordering out of the box
cheers

**Comment by niebert** [comment]
Thank you Spencer, excellent, try to embed and integrate in [https://niebert.github.io/Wiki2Reveal Wiki2Reveal] cheers

**Comment by fractalien** [comment]
I'm sorry in case I'm missing something obvious, but how would I get the "better ordering out of the box"? Calling .json() on a section in 7.2.9, I still get lists after paragraphs, not in them. (Trying with [this article](https://en.wikipedia.org/wiki/Mercury).) Is there an accessor method to get both sentences and lists from a paragraph object?

**Comment by spencermountain** [comment]
hi @fractalien no you're right, this is still not possible.

**Comment by fractalien** [comment]
Thanks! I still got to work around my problem, namely that the [article](https://en.wikipedia.org/wiki/Mercury)'s API-sourced [markup](https://en.wikipedia.org/w/api.php?format=json&action=query&prop=extracts&exsectionformat=wiki&exlimit=1&explaintext&redirects=true&formatversion=2&titles=Mercury) fails to put a newline after the list ([bug](https://phabricator.wikimedia.org/T208132)), yielding text containing "SunMercury". Given that the newer [REST API](https://en.wikipedia.org/api/rest_v1/#/) seems not to offer a full-article markup output, I was between a rock and a hard place:
1. Applying `Wtf(…).json()` to the erroneous API result obviously didn't magically un-break the text.
2. When using `Wtf.fetch(…).json()`, I got the lists separated from sentences, out of order.

My solution was to apply Wtf twice, essentially `Wtf(Wtf.fetch(…).markdown()).json()` to get proper parsing, but no separation between sentences and lists. Frankly, I am surprised that there are different json() results depending on whether Wtf does the fetching itself or not, but perhaps I missed the existence of a parameter – I'll take it as a feature, anyway. :)

**Comment by spencermountain** [comment]
ha! this is very clever.
i wish things were easier!
